cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_test.go

```diff
@@ -3004,6 +3006,207 @@ func TestDeleteInstalledPackage(t *testing.T) {
+					},
+					Type: "Opaque",
+					Data: map[string][]byte{
+						"values.yaml": []byte("foo: bar"),
+					},
+				},
+				&k8scorev1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "default",
+						Name:      "my-installation-ctrl",
+					},
+					Data: map[string]string{
+						"spec": "{\"labelKey\":\"kapp.k14s.io/app\",\"labelValue\":\"my-id\"}",
+					},
+				},
+				&k8scorev1.Pod{
+					TypeMeta: metav1.TypeMeta{
+						APIVersion: "v1",
+						Kind:       "Pod",
+					},
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "default",
+						Name:      "my-installation-pod",
+						Labels:    map[string]string{"kapp.k14s.io/app": "my-id"},
+					},
+					Spec: k8scorev1.PodSpec{
+						Containers: []k8scorev1.Container{{
+							Name: "my-installation-container",
+						}},
+					},
+				},
+			},
+			expectedStatusCode: codes.OK,
+			expectedResponse: &corev1.GetInstalledPackageResourceRefsResponse{
+				ResourceRefs: []*corev1.ResourceRef{
+					{
+						ApiVersion: "v1",
+						Kind:       "Pod",
+						Name:       "my-installation-pod",
+						Namespace:  "default",
+					},
+				},
+				Context: defaultContext,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var unstructuredObjects []runtime.Object
+			for _, obj := range tc.existingObjects {
+				unstructuredContent, _ := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+				unstructuredObjects = append(unstructuredObjects, &unstructured.Unstructured{Object: unstructuredContent})
+			}
+
+			typedClient := typfake.NewSimpleClientset(tc.existingTypedObjects...)
+			dynClient := dynfake.NewSimpleDynamicClientWithCustomListKinds(
+				runtime.NewScheme(),
+				map[schema.GroupVersionResource]string{
+					{Group: datapackagingv1alpha1.SchemeGroupVersion.Group, Version: datapackagingv1alpha1.SchemeGroupVersion.Version, Resource: pkgsResource}:         pkgResource + "List",
+					{Group: datapackagingv1alpha1.SchemeGroupVersion.Group, Version: datapackagingv1alpha1.SchemeGroupVersion.Version, Resource: pkgMetadatasResource}: pkgMetadataResource + "List",
+					{Group: packagingv1alpha1.SchemeGroupVersion.Group, Version: packagingv1alpha1.SchemeGroupVersion.Version, Resource: pkgInstallsResource}:          pkgInstallResource + "List",
+					{Group: "", Version: "v1", Resource: "pods"}: "Pod" + "List",
+				},
+				unstructuredObjects...,
+			)
+
+			s := Server{
+				clientGetter: func(context.Context, string) (kubernetes.Interface, dynamic.Interface, error) {
+					return typedClient, dynClient, nil
+				},
+				kappClientsGetter: func(ctx context.Context, cluster, namespace string) (ctlapp.Apps, ctlres.IdentifiedResources, *kappcmdapp.FailingAPIServicesPolicy, ctlres.ResourceFilter, error) {
+					// Create a fake DepsFactory and pass it the fake k8s clients
+					depsFactory := NewFakeDepsFactoryImpl()
+					depsFactory.SetCoreClient(typedClient)
+					depsFactory.SetDynamicClient(dynClient)
```

Comment:
This is the approach I've been able to come up with for faking the kapp internal clients. Even if it _kinda_ works, it returns an empty list of resources found :(
If you have any suggestions or can give it a try, I'd really appreciate it!

Reply:
So I setup the vscode debugger (which was pretty simple - hadn't used it much before) and traced this down from your call to `	resources, err := resourcesClient.List(labelSelector, nil)` in `server_data_resources.go`. From there:

- kapp first tries to [list all the resource types](https://github.com/vmware-tanzu/carvel-kapp/blob/v0.44.0/pkg/kapp/resources/identified_resources_list.go#L17) on the cluster, which, for your test client, results in an empty slice, so that
- kapp [goes on to call `resources.All(resTypes, allOpts)`](https://github.com/vmware-tanzu/carvel-kapp/blob/v0.44.0/pkg/kapp/resources/identified_resources_list.go#L45), but this just iterates the slice of `resTypes`, so also returns an empty slice.

Looking at the implementation, the core issue is the [function which collects the resource types](https://github.com/vmware-tanzu/carvel-kapp/blob/v0.44.0/pkg/kapp/resources/resource_types.go#L111) is using the discovery api, so in your test it's not returning any APIs. Might be worth re-checking the kapp test code to see how they setup the discovery API in tests or otherwise ensure certain resource types (pods, in your case) are returned.

HTH!